blockchain/upgrade.go

```diff
@@ -3363,6 +3428,443 @@ func separateUtxoDatabase(ctx context.Context, db database.DB, utxoDb database.D
+	bucketId, err = fetchLegacyBucketId(utxoBackend.Get, utxoSetLegacyBucketName)
+	if err != nil {
+		return err
+	}
+
+	// If the legacy bucket doesn't exist, return as there is nothing to do.
+	if bucketId == nil {
+		return nil
+	}
+
+	// doBatch contains the primary logic for migrating the UTXO set.  This is
+	// done because attempting to migrate in a single database transaction could
+	// result in massive memory usage and could potentially crash on many systems
+	// due to ulimits.
+	//
+	// It returns whether or not all entries have been updated.
+	const maxEntries = 20000
+	var totalMigrated uint64
+	doBatch := func(tx UtxoBackendTx) (bool, error) {
+		// Migrate UTXO set entries so long as the max number of entries for this
+		// batch has not been exceeded.
+		var logProgress bool
+		var numMigrated uint32
+
+		iter := tx.NewIterator(bucketId)
+		defer iter.Release()
+		for iter.Next() {
+			// Reset err on each iteration.
+			err = nil
+
+			if interruptRequested(ctx) {
+				logProgress = true
+				err = errInterruptRequested
+				break
+			}
+
+			if numMigrated >= maxEntries {
+				logProgress = true
+				err = errBatchFinished
+				break
+			}
+
+			// Move the UTXO set entry to the new key.
+			oldKey := iter.Key()
+			newKey := prefixedKey(utxoPrefixUtxoSetV3, oldKey[len(bucketId):])
+			err = moveKey(tx, oldKey, newKey)
+			if err != nil {
+				return false, fmt.Errorf("error migrating UTXO set entry: %v", err)
+			}
+
+			numMigrated++
+		}
+		if iterErr := iter.Error(); iterErr != nil {
+			return false, convertLdbErr(iterErr, iterErr.Error())
+		}
+		isFullyDone := err == nil
+		if (isFullyDone || logProgress) && numMigrated > 0 {
+			totalMigrated += uint64(numMigrated)
+			log.Infof("Migrated %d entries (%d total)", numMigrated,
+				totalMigrated)
+		}
+		return isFullyDone, err
+	}
+
+	// Migrate all entries in batches for the reasons mentioned above.
+	if err := utxoBackendBatchedUpdate(ctx, utxoBackend, doBatch); err != nil {
+		return err
+	}
+
+	elapsed := time.Since(start).Round(time.Millisecond)
+	log.Infof("Done migrating UTXO database.  Total entries: %d in %v",
+		totalMigrated, elapsed)
+
+	return nil
+}
+
+// dbPutUtxoBackendInfoV1 uses an existing UTXO backend transaction to store the
+// V1 UTXO backend information.
+func dbPutUtxoBackendInfoV1(tx UtxoBackendTx, info *UtxoBackendInfo) error {
+	// V1 database info keys.
+	utxoDbInfoVersionKeyNameV1 := []byte("version")
+	utxoDbInfoCompVerKeyNameV1 := []byte("compver")
+	utxoDbInfoUtxoVerKeyNameV1 := []byte("utxover")
+	utxoDbInfoCreatedKeyNameV1 := []byte("created")
+
+	// V1 bucket info.
+	bucketNameV1 := []byte("dbinfo")
+	bucketIdV1, err := fetchLegacyBucketId(tx.Get, bucketNameV1)
+	if err != nil {
+		return err
+	}
+
+	// uint32Bytes is a helper function to convert a uint32 to a byte slice
+	// using the byte order specified by the database namespace.
+	uint32Bytes := func(ui32 uint32) []byte {
+		var b [4]byte
+		byteOrder.PutUint32(b[:], ui32)
```

Comment:
This and other references to `byteOrder` are using the global outside of the upgrade code that might change over time.  Granted, it probably won't change over time, but it should define it locally as is done in the `deserializeDatabaseInfoV2` func in this file to be safe.

Reply:
Fair point, I updated `byteOrder` to be defined locally in that function.